Add initial VC API Exchanges docs. (#50)
credential-handler/chapi.io: developers/wallets/exchanges.md

---
layout: subpage
title: "Handling a VC API Exchange in a Web Wallet"
permalink: /developers/wallets/exchanges/
---

## {{ title }}

> A VC API workflow defines a particular set of steps for exchanging verifiable
> credentials between two parties across a trust boundary. Each step can involve
> the issuance, verification, transmission, or presentation of verifiable
> credentials.

A Web Wallet may receive an Exchange URL through CHAPI, reading a QR code, or
via some other user initiated transmission.

Once that URL is received, the Wallet can initiate the exchange (acting as the
_exchange client_) by sending a POST request to that URL.

Exchange URL:
```

> https://vcapi.example.com/exchanges/12345/

Reviewer: ```suggestion
https://vcapi.example.com/workflows/abcdef/exchanges/12345
```

Author: Fixed in 1c6994b

```

To initiate an exchange using VC API, an exchange client performs an HTTP POST
sending a JSON object as the request body. In the simplest case, when the client
has no constraints of its own on the exchange — i.e., it has nothing to request
from the other party — the JSON object is empty ({})--as seen below:

```http

> POST /exchanges/12345/

Reviewer: ```suggestion
POST /workflows/abcdef/exchanges/12345
```

Author: Fixed in 1c6994b

        "example": {
          "@context": [
            "https://www.w3.org/ns/credentials/v2",
            "https://w3id.org/citizenship/v1"
          ],
          "type": "PermanentResidentCard"
        }
      }]
    }]
  }
}
```

If the object includes `verifiablePresentation`, then some *information is
offered*, such as verifiable credentials issued to the holder operating the
exchange client (i.e. the Wallet) or verifiable credentials with information
about the exchange server's operator based on the exchange client's request.

For example:
```json
{
  "verifiablePresentation": {
    "@context": [
      "https://www.w3.org/ns/credentials/v2"
    ],
    "type": ["VerifiablePresentation"],
    "verifiableCredential": [{
      "@context": [
        "https://www.w3.org/ns/credentials/v2",
        "https://w3id.org/citizenship/v1"
      ],
      "type": ["VerifiableCredential", "PermanentResidentCard"],
      // additional properties...
    }]
  }
}
```

If the object includes `redirectUrl`, the *exchange is complete* and the
workflow service recommends that the client proceed to another place to continue
the interaction in another form.

For example:
```json
{

>   "redirectUrl": "https://vcapi.example.com/go-here-next/"

Reviewer: ```suggestion
  "redirectUrl": "https://vcapi.example.com/go-here-next"
```

Below is an example of a typical exchange:

```mermaid
sequenceDiagram
    participant H as Holder
    participant W as Holder Coordinator (Wallet)
    participant I as Issuer/Verifier Coordinator
    autonumber
    Note right of H: Start exchange
    W->>I: Initiate
    Note right of W: POST /workflows/123/exchanges/123 &mdash; HTTP request to start exchange (e.g., send credentials, get credentials)
    I->>W: Verifiable Presentation Request (VPR)
    Note left of I: VPR includes method of interaction, for purposes of exchange
    W->>I: Verifiable Presentation (VP)
    Note right of W: POST /workflows/123/exchanges/abc &mdash; sent via interaction mechanism to meet requirements of exchange
    I->>W: Verifiable Presentation
    Note left of I: VP includes result of exchange (e.g., VCs), or VPR with new interaction request, or error description
```

The exchange client (Wallet) code for a flow like the above may look similar to
the following:

```js
const receivedExchangeUrl = 'https://vcapi.example.com/workflows/123/exchanges/123';

const response = await fetch(receivedExchangeUrl, {
  method: 'POST'
  body: JSON.stringify({})
});

> const body = response.json();

Reviewer: I can't remember the fetch API ... is this needed?

```suggestion
const body = await response.json();
```

function checkResponse(body) {
  if('verifiablePresentationRequest' in body) {
    // use the information in the Verifiable Presentation Request to find a

>     // credentail that fullfills the request, then send a Verifiable

Reviewer: ```suggestion
    // credential that fulfills the request, then send a Verifiable
```

    // Presentation as the response to the interaction endpoint included in the
    // Verifiable Presentation Request

>     const presentation = findCredentialAndCreatePresentation();

Reviewer: ```suggestion
    const verifiablePresentation = findCredentialAndCreatePresentation();
```

    // TODO: properly extract interaction URL
    const interactionUrl = body.interact.service[0].serviceEndpoint;
    const response = await fetch(interactionUrl, {
      method: 'POST',
      body: JSON.stringify(presentation)
    });
    checkResponse(response.json());
  }

>   // TODO: should these be exclusive?

Reviewer: No, they can occur together.

Author: Fixed in c723a31

  if('verifiablePresentation' in body) {
    // the Wallet has received a Verifiable Presentation containing one or more

>     // Verifiable Credentials--validate and verify them per your use case

Reviewer: ```suggestion
    // Verifiable Credentials--use them per your use case
```

  }
  if('redirectUrl' in body) {

>     // TODO: take the user to the new location OR does the exchange "move"?

Reviewer: Exchange does not "move", the user is just taken elsewhere for any other kind of continuing interaction with the coordinator.

```suggestion
    // take the user to the new location
```

---
layout: subpage
title: "Handling a VC API Exchange in a Web Wallet"
permalink: /developers/wallets/exchanges/
---

## {{ title }}

> A VC API workflow defines a particular set of steps for exchanging verifiable
> credentials between two parties across a trust boundary. Each step can involve
> the issuance, verification, transmission, or presentation of verifiable

> > credentials.

Reviewer: Can we link to the spec's "Workflows and Exchanges" section here?

Author: Done in https://github.com/credential-handler/chapi.io/commit/6c51f3b4c80fe917ad139653b78a0d1b9847b8e8

A VC API workflow interaction begins with _either_ a CHAPI event or an
interaction URL via a QR code (or similar user initiated transfer experience).

The CHAPI event will provide a complete protocols object similar to the
following:
```json
{
  "protocols": {
    "vcapi": "https://vcapi.example.com/workflows/abc/exchanges/12345",
    "OIC4VCI": "openid-credential-offer://?..."
  }
}
```
Alternatively, a QR code can be used to provide a URL which when dereferenced
will result in either a `protocols` (object as above) or an HTML fallback page
to allow the user to continue otherwise.

To retrieve a `protocols` object from an _interaction URL_, the Wallet must send
an HTTP GET request including an explicit `Accept: application/json` request
header--which results in the same JSON object as above:

```http

> GET /exchanges/12345/protocols

Reviewer: ```suggestion
GET /workflows/abc/exchanges/123/protocols
```

Author: Fixed in bce5fc3

Host: vcapi.example.com
Accept: application/json

{
  "protocols": {

>     "vcapi": "https://vcapi.example.com/exchanges/12345",

Reviewer: ```suggestion
    "vcapi": "https://vcapi.example.com/workflows/abc/exchanges/123",
```

Author: Fixed in bce5fc3

    "OIC4VCI": "openid-credential-offer://?..."
  }
}
```

Once a VC API exchange URL is acquisitioned from `protocols.vcapi`, a POST
request is sent with a configuration object (which may be empty) to begin the
exchange:

```http

> POST /exchanges/12345

Reviewer: ```suggestion
POST /workflows/abc/exchanges/123
```

Author: Fixed in bce5fc3

Host: vcapi.example.com

{}
```

A response will be returned by the exchanger...

If the response object is empty (as above), the exchange is complete and nothing
is requested from nor offered to the exchange client.

> If, however, the object includes `verifiablePresentationRequest`, then the

Reviewer 2: Should this link to the [vp-request spec](https://w3c-ccg.github.io/vp-request-spec/#overview)?

Reviewer 2: @BigBlueHat I also wonder if it would be helpful to have a follow-up PR with "FAQ" that can help direct newcomers unfamiliar with VC to the relevant specs. I know I would want more information on how to write or parse these objects.

Author: Yeah...this whole site needs reworking into proper documentation...

Author: > Should this link to the [vp-request spec](https://w3c-ccg.github.io/vp-request-spec/#overview)?

Fixed in a7bf93d.

The FAQ will have to wait, but I agree it's a needed addition!

exchange is not yet complete and some *additional information is requested*, as
specified by the contents of the associated verifiable presentation request.

For example:
```json
{
  "verifiablePresentationRequest": {
    "query": [{
      "type": "QueryByExample",
      "credentialQuery": [{

>         "reason": "Please present proof of citizenship.",

Reviewer: Trying to make this more of a "reason" without being too specific:

```suggestion
        "reason": "We need proof of citizenship to onboard you.",
```
